README.md
# bootprint-json-schema 

[![NPM version](https://img.shields.io/npm/v/bootprint-json-schema.svg)](https://npmjs.com/package/bootprint-json-schema)
[![Travis Build Status](https://travis-ci.org/bootprint/bootprint-json-schema.svg?branch=master)](https://travis-ci.org/bootprint/bootprint-json-schema)
[![Coverage Status](https://img.shields.io/codecov/bootprint/bootprint-json-schema.svg)](https://codecov.io/github/bootprint/bootprint-json-schema)
[![Greenkeeper badge](https://badges.greenkeeper.io/bootprint/bootprint-json-schema.svg)](https://greenkeeper.io/)

> Converts a json-schema into a static html page

This bootprint-module can be used to convert a json-schema to a human-readable HTML.

## Features

* Complete support for the keywords from JSON-schema [draft-04](https://tools.ietf.org/html/draft-fge-json-schema-validation-00), [draft-05](https://tools.ietf.org/html/draft-wright-json-schema-validation-00) and [draft-06](https://tools.ietf.org/html/draft-wright-json-schema-validation-01)
* Auto-detection of the schema-version based on the `$schema`-property.
* Auto-detection can be overridden by setting e.g. `"$explicitSchemaVersion": "draft-05"`
* Links to the specification for all keywords

* Full test coverage of JavaScript-code
* Lots of tests to cover Handlebars templates and partials


# Installation

```
npm -g install bootprint
npm -g install bootprint-json-schema
```

## Usage


After installing the package globally, you can run bootprint with the command

```bash
bootprint json-schema https://raw.githubusercontent.com/bootprint/bootprint-json-schema/v2.0.0-rc.3/examples/example.json target
```

`example.json` can be found in [examples/example.json](examples/example.json) in this project.
It is a copy of the [fstab-example](http://json-schema.org/example2.html) from the JSON-schema documentation, licensed under the AFL or BSD license.


Bootprint will then generate the following files:

<pre><code>examples/target/
├── index.html
├── main.css
└── main.css.map
</code></pre> 


## Customizing

You can write your own module that customizes the partials and helpers in this module
(see [the bootprint documentation](https://github.com/bootprint/bootprint/blob/master/doc/modules.md)) for details.

The entrypoint JavaScript-file of such a module would look like.

```js
module.exports = function (customize) {
  return customize
    .load(require('bootprint-json-schema'))
    .merge({
      // You customizations here
    })

// Add "package" metadata. This can be evaluated by documentation generators
module.exports.package = require('./package')
```

# API

see [docs/api.md](docs/api.md)


# License

`bootprint-json-schema` is published under the MIT-license.

See [LICENSE](LICENSE) for details.


# Release-Notes
 
For release notes, see [CHANGELOG.md](CHANGELOG.md)
 
# Contributing guidelines

See [CONTRIBUTING.md](CONTRIBUTING.md).

CHANGELOG.md
# Release notes


<a name="current-release"></a>
# Version 2.0.0-rc.3 (Sun, 06 Aug 2017 13:16:33 GMT)

* [d40ed07](https://github.com/bootprint/bootprint-json-schema/commit/d40ed07) Update documentation - Nils Knappmeier
* [3192776](https://github.com/bootprint/bootprint-json-schema/commit/3192776) Add tests for multipleOf for draft-04 and -06 - Nils Knappmeier
* [cadc7e6](https://github.com/bootprint/bootprint-json-schema/commit/cadc7e6) Support for JSON-schema draft-05 - Nils Knappmeier

# Version 2.0.0-rc.2 (Sat, 05 Aug 2017 22:41:02 GMT)

* [955e501](https://github.com/bootprint/bootprint-json-schema/commit/955e501) Update documentation - Nils Knappmeier
* [c947822](https://github.com/bootprint/bootprint-json-schema/commit/c947822) Update documentation by running "npm run thought" - Nils Knappmeier
* [8040014](https://github.com/bootprint/bootprint-json-schema/commit/8040014) BREAKING: Complete rewrite of all partials, templates and helpers - Nils Knappmeier

# Version 2.0.0-rc.1 (Sun, 25 Jun 2017 22:01:12 GMT)

* [01e6d2b](https://github.com/bootprint/bootprint-json-schema/commit/01e6d2b) Bump version of bootprint-base to get better helper docs - Nils Knappmeier
* [3867ef7](https://github.com/bootprint/bootprint-json-schema/commit/3867ef7) Measures to increase test-coverage - Nils Knappmeier
* [1b5605f](https://github.com/bootprint/bootprint-json-schema/commit/1b5605f) BREAKING: Remove unused helpers 'json_schema__subschema_name' and 'json_schema__resolve_ref' - Nils Knappmeier
* [287afa2](https://github.com/bootprint/bootprint-json-schema/commit/287afa2) Remove empty obsolete test-file - Nils Knappmeier
* [814d86a](https://github.com/bootprint/bootprint-json-schema/commit/814d86a) BREAKING: Refactor helper names to be valid JavaScript function names - Nils Knappmeier
* [038a06b](https://github.com/bootprint/bootprint-json-schema/commit/038a06b) Bump dependency versions and adapt, update docs and chore - Nils Knappmeier
* [9f42f0e](https://github.com/bootprint/bootprint-json-schema/commit/9f42f0e) Prepare next version - Nils Knappmeier
* [0225c10](https://github.com/bootprint/bootprint-json-schema/commit/0225c10) Update dependencies to enable Greenkeeper 🌴 (#11) - greenkeeper[bot]

# Version 1.1.0 (Mon, 13 Feb 2017 21:33:41 GMT)

* [c83458c](https://github.com/bootprint/bootprint-json-schema/commit/c83458c) Add string pattern (#10) - Alex Collins

# Version 1.0.0 (Sat, 24 Dec 2016 20:43:54 GMT)

* [bb00a2e](https://github.com/bootprint/bootprint-json-schema/commit/bb00a2e) Prepare for 1.0 release (#9) - Nils Knappmeier

# Version 0.8.6 (Thu, 24 Mar 2016 07:31:16 GMT)

* [4d70db8](https://github.com/bootprint/bootprint-json-schema/commit/4d70db8) #3: Corrected typo in partial (maxLength-variable spelled wrongly) - Nils Knappmeier

# Version 0.8.5 (Tue, 15 Mar 2016 14:27:33 GMT)

* [40ca080](https://github.com/bootprint/bootprint-json-schema/commit/40ca080) Generate README with "thought" - Nils Knappmeier
* [2222c27](https://github.com/bootprint/bootprint-json-schema/commit/2222c27) Add "require" to main-spec in order to get (low) coverage - Nils Knappmeier
* [2cfccb8](https://github.com/bootprint/bootprint-json-schema/commit/2cfccb8) Add mocha and chai as dev-dependencies - Nils Knappmeier
* [98a14f8](https://github.com/bootprint/bootprint-json-schema/commit/98a14f8) Adjust travis-configuration - Nils Knappmeier
* [3b7b322](https://github.com/bootprint/bootprint-json-schema/commit/3b7b322) Move to bootprint-organization and enable ghook for StandardJS - Nils Knappmeier

## v0.8.4 - 2015-10-21

# Fix

* Remove `bootprint` as peer-dependency, because `npm` versions 1 and 2 download it needlessly.

## v0.8.3 - 2015-10-19

# Fix

* Typo in `files`-property, causing the build to be broken (Thanks @davoad for PR #6)

## v0.8.2 - 2015-10-19 (BROKEN)

# Fix

* Add `files`-property to package.json

## v0.8.1 - 2015-10-17

### Add

* Enums are linked to subschema-definitions, if a property is a swagger-discriminator

## v0.8.0 - 2015-10-14

### Change

* bootprint-swagger#39: Attempting to resolve invalid JSON-references (e.g. for Swagger responses) 
  now throws an error

## v0.7.2 - 2015-10-13 

### Add

* Enums are linked to subschema-definitions, if a property is a swagger-discriminator 

### Fix

* Silently ignore missing `items` tag in `array`-definitions

## v0.7.1 - 2015-09-27
### Fix

* bootprint-swagger#35: "allOf" renders incorrect HTML doc

## v0.7.0 - 2015-09-25
### Add

* Display string-length limits

## v0.6.2 - 2015-09-22
### Fix 

* Be forgiving on calling datatype-helper with empty argument

## v0.6.1 - 2015-09-19
### Fix

* Helper for resolving $ref-properties now returns "undefined" instead of an exception
  when a path cannot be followed completely (because a property in the middle of the path 
  is missing)

## v0.6.0 - 2015-09-18
### Add

* Helper for resolving $ref-properties to their reference object (needed for bootprint-swagger#31)

## v0.5.3 - 2015-09-10
### Fix

* bootprint-swagger#28: Read-only fields in parameter schema

## v0.5.2 - 2015-09-05
### Fix

* bootprint-swagger#26: Definition objects not getting any details

## v0.5.1 - 2015-09-01
### Fix

* bootprint-swagger#18: Details about primitive (string, number, etc). array-item-types were not displayed.

## The dark past

* You'll have to look through the commit log to get details.


test/utils.js
module.exports = {
  select
}

/**
 * Returns a cheerio-selector for the an element within the section of a definition
 * @param {string} definition key of the definitions-object
 * @param {string} keyword the keyword section to check
 * @param {string} selector a selector for elements within the section
 * @returns {string}
 */
function select (definition, keyword, selector) {
  selector = selector || ''
  return `#definitions-${definition} > .panel-body > .json-schema--schema > section.json-schema--section[data-keywords*="${keyword}"] ${selector}`
}


test/helper-spec.js
/* global describe */
/* global it */

var expect = require('chai').expect

var helpers = require('../handlebars/helpers.js')
var Handlebars = require('handlebars').create()
Handlebars.registerHelper(helpers)

function run (template, json) {
  return Handlebars.compile(template)(json)
}

describe('The Handlebars-helpers:', function () {
  describe('the json_schema__datatype helper', function () {
    // Call the datatype-helper
    function dt (obj) {
      return run('{{json_schema__datatype type}}', {type: obj})
    }

    it('should handle undefined gracefully', function () {
      expect(dt(undefined)).to.equal('')
    })

    it('should handle null gracefully', function () {
      expect(dt(null)).to.equal('')
    })

    it('should return an empty string for null and undefined gracefully, as well as composite schema, since they are rendered in  partials ', function () {
      expect(dt({anyOf: []})).to.equal('')
      expect(dt({oneOf: []})).to.equal('')
      expect(dt({allOf: []})).to.equal('')
    })

    it('should treat unspecific types as object', function () {
      expect(dt({})).to.equal('object')
    })

    it('should treat unspecific array types as object[]', function () {
      expect(dt({type: 'array'})).to.equal('object[]')
    })

    it('should treat arrays with unspecific types object[]', function () {
      expect(dt({type: 'array', items: {}})).to.equal('object[]')
    })

    it('should write type[] for specific arrays', function () {
      expect(dt({type: 'array', items: {type: 'string'}})).to.equal('string[]')
    })

    it('should write recurse multiple steps of array types', function () {
      expect(dt({type: 'array', items: {type: 'array', items: {type: 'string'}}})).to.equal('string[][]')
    })
  })

  describe('the json_schema__range helper', function () {
    // Call the datatype-helper
    function range (obj) {
      return run('{{{json_schema__range range}}}', {range: obj})
    }

    it('should handle empty ranges gracefully', function () {
      expect(range({})).to.equal('')
    })

    it('should handle empty integer ranges gracefully', function () {
      expect(range({ type: 'integer' })).to.equal('')
    })

    it('should handle empty number ranges gracefully', function () {
      expect(range({ type: 'number' })).to.equal('')
    })

    it('should ignore minimum and maximum for non-numeric types', function () {
      expect(range({type: 'string', minimum: 2, maximum: 3})).to.equal('')
    })

    it('should render a range for only minimum (starting with coma)', function () {
      expect(range({minimum: 2, type: 'number'})).to.equal(', { x ∈ ℝ | x ≥ 2 }')
    })

    it('should render a range for only maximum (starting with coma)', function () {
      expect(range({maximum: 2, type: 'number'})).to.equal(', { x ∈ ℝ | x ≤ 2 }')
    })

    it('should render a range with minimum and maximum (starting with coma)', function () {
      expect(range({maximum: 2, minimum: 0, type: 'number'})).to.equal(', { x ∈ ℝ | 0 ≤ x ≤ 2 }')
    })

    it('should render a open range with minimumExclusive (starting with coma)', function () {
      expect(range({minimum: 0, minimumExclusive: true, type: 'number'})).to.equal(', { x ∈ ℝ | x > 0 }')
    })

    it('should render a open range with maximumExclusive (starting with coma)', function () {
      expect(range({maximum: 2, maximumExclusive: true, type: 'number'})).to.equal(', { x ∈ ℝ | x < 2 }')
    })

    it('should render a range with minimumExclusive and maximum (starting with coma)', function () {
      expect(range({
        maximum: 2,
        minimum: 0,
        minimumExclusive: true,
        type: 'number'
      })).to.equal(', { x ∈ ℝ | 0 < x ≤ 2 }')
    })

    it('should render a range with minimum and maximumExclusive (starting with coma)', function () {
      expect(range({
        maximum: 2,
        maximumExclusive: true,
        minimum: 0,
        type: 'number'
      })).to.equal(', { x ∈ ℝ | 0 ≤ x < 2 }')
    })

    it('should render the correct set-symbol for integers', function () {
      expect(range({type: 'integer', minimum: 0})).to.equal(', { x ∈ ℤ | x ≥ 0 }')
    })
  })
})


test/fstab/fstab-spec.js
/*!
 * bootprint-swagger <https://github.com/nknapp/bootprint-swagger>
 *
 * Copyright (c) 2015 Nils Knappmeier.
 * Released under the MIT license.
 */

/* global describe */
/* global it */
/* global before */
var expect = require('chai').expect

var tester = require('bootprint-unit-testing')

describe('The fstab-fixture', function () {
  this.timeout(10000)
  var bptest = tester(require('../..'), __dirname, require('./schema.json'))

  // Run bootprint. The parsed "index.html"-file (cheerio) is then available
  // under "bptest.$"
  before(bptest.run)

  it('should have a pattern-block in the fstab-section', function () {
    expect(bptest.$('#definition-diskUUID .json-property-pattern .json-schema--regex').text())
      .to.equal('^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$')
  })
})


handlebars/helpers.js
var util = require('util')

/* eslint-disable camelcase */

/**
 * Handlebars helpers for `bootprint-json-schema`
 * @name helpers
 */
module.exports = {
  json_schema__datatype,
  json_schema__range
}

/**
 * Returns a descriptive string for a datatype
 * @param {object} value a json-schema datatype-object
 * @returns {String} a string like <code>string[]</code> or <code>object[][]</code>
 * @access public
 * @memberOf helpers
 */
function json_schema__datatype (value) {
  if (!value) return null
  if (value['anyOf'] || value['allOf'] || value['oneOf']) {
    return ''
  }
  if (!value.type) {
    return 'object'
  }
  if (value.type === 'array') {
    return json_schema__datatype(value.items || {}) + '[]'
  }
  return value.type
}

/**
 *
 * @param range a json-schema object with minimum, maximum, exclusiveMinimum, exclusiveMaximum
 * @param {number=} [range.minimum]
 * @param {number=} [range.maximum]
 * @param {string} [range.type] the json-type (integer, or number)
 * @param {boolean} [range.minimumExclusive]
 * @param {boolean} [range.maximumExclusive]
 * @access public
 * @memberOf helpers
 */
function json_schema__range (range) {
  switch (range.type) {
    case 'integer':
    case 'number':
      return numericRange(range)
    case 'string':
      return stringLengthRange(range)
    default:
      return ''
  }
}

/**
 * Render a range for numeric types (integer, number)
 * @param range
 * @returns {*}
 */
function numericRange (range) {
  const hasMinimum = range.minimum || range.minimum === 0
  const hasMaximum = range.maximum || range.maximum === 0

  if (!hasMinimum && !hasMaximum) {
    // There is no range
    return ''
  }

  let numberSet = range.type === 'integer'
    ? '\u2208 \u2124' // ELEMENT OF - DOUBLE-STRUCK CAPITAL Z
    : '\u2208 \u211D' // ELEMENT OF - DOUBLE-STRUCK CAPITAL R

  if (hasMinimum && !hasMaximum) {
    return util.format(', { x %s | x %s %d }',
      numberSet,
      range.minimumExclusive ? '>' : '\u2265',
      range.minimum)
  } else if (hasMaximum && !hasMinimum) {
    return util.format(', { x %s | x %s %d }',
      numberSet,
      range.maximumExclusive ? '<' : '\u2264',
      range.maximum)
  } else {
    // if (hasMaxmium && hasMinimum)
    return util.format(', { x %s | %d %s x %s %d }',
      numberSet,
      range.minimum,
      range.minimumExclusive ? '<' : '\u2264',
      range.maximumExclusive ? '<' : '\u2264',
      range.maximum)
  }
}

/**
 * Render string length restrictions
 * @param range
 * @param {number} range.maxLength
 * @param {number} range.minLength
 */
function stringLengthRange (range) {
  if (range.maxLength && range.minLength) {
    return `, ${range.minLength} to ${range.maxLength} chars`
  }
  if (range.maxLength) {
    return `, up to ${range.maxLength} chars`
  }
  if (range.minLength) {
    return `, at least ${range.minLength} chars`
  }
  return ''
}


handlebars/helpers/index.js
/* eslint-disable camelcase */

/**
 * Default Handlebars-helpers for `bootprint-json-schema`
 * @name helpers
 */
module.exports = {
  json_schema__doclink,
  json_schema__datatype,
  json_schema__number_range,
  json_schema__subschema_name,
  json_schema__definition_id,
  json_schema__could_be_numeric,
  json_schema__could_be_of_type,
  json_schema__is_array,
  json_schema__split_coma,
  json_schema__is_required,
  json_schema__has_any,
  json_schema__enumerate,
  json_schema__count_range,
  json_schema__if_version
}

const {htmlId} = require('bootprint-base/handlebars/helpers')
const {safe} = require('./utils')

const versionDispatch = {
  'draft-04': require('./helpers-draft-04'),
  'draft-05': require('./helpers-draft-05'),
  'draft-06': require('./helpers-draft-06')
}

const schemaByUrl = {
  'http://json-schema.org/draft-04/schema#': 'draft-04',
  'http://json-schema.org/draft-06/schema#': 'draft-06'
}

/**
 * Render a link to a json-schema docs section from the json-schema documentation.
 * If the section does not exist, it is assumed to be a vendor extension.
 * The mapping of names to section numbers can be found in the version-specific helper
 * files.
 *
 * @param {string} type the type of section (keywords, formats)
 * @param {string} sectionName the section name (e.g. items)
 * @param {object} options the Handlebars options
 * @return {string|Handlebars.SafeString} a SafeString containing a link to the documentation or
 *   'vender extension' if there is no section of the given name
 * @access public
 * @memberOf helpers
 */
function json_schema__doclink (type, sectionName, options) {
  let version = determineVersion(options)
  let {docLink, sections} = versionDispatch[version].metadata

  let section = sections[type][sectionName]
  if (section) {
    return safe`<a href="${docLink}#section-${section}">${section}</a>`
  } else {
    return 'vendor extension'
  }
}

/**
 * Executes the helper-block, if the version matches any of the
 * specified target versions
 * @param {...string} targetVersions a list of schema-versions to match against (varargs)
 * @access public
 * @memberOf helpers
 */
function json_schema__if_version (...targetVersions /*, options */) {
  const options = targetVersions[targetVersions.length - 1]
  targetVersions = targetVersions.slice(0, targetVersions.length - 1)
  const schemaVersion = determineVersion(options)
  if (targetVersions.indexOf(schemaVersion) >= 0) {
    return options.fn(this)
  } else {
    return options.inverse(this)
  }
}

/**
 * Returns the datatype of a schema as short formal string like `array<string|integer>`
 *
 * @param {object} schema a json-schema
 * @param {string|string[]} types one or more types of potentially multiple types in the schema that should be used right now.
 * @returns {String} a string like `array<string|integer>`
 * @access public
 * @memberOf helpers
 */
function json_schema__datatype (schema, types) {
  if (schema == null) {
    return null
  }
  /**
   * The type(s) that are processed
   * @type {string|string[]}
   */
  if (types == null) {
    return '*'
  }
  if (Array.isArray(types)) {
    // "type" is an array (i.e. multiple types) -> process each type
    return types.map(type => json_schema__datatype(schema, type)).join('|')
  }
  // "type" is a string
  if (types === 'array') {
    let items = schema.items || {}
    return `array<${json_schema__datatype(items, items.type)}>`
  }
  return types
}

/**
 * Return a string for the numeric range restriction of the given schema
 * (as determined by `minimum`, `maximum`, `exclusiveMinimum` and `exclusiveMaximum`
 * properties).
 *
 * The implementation is dispatched to version-specific behavior because `draft-04`
 * and `draft-05` differ from `draft-06`
 * @param {object} schema the current (sub-)schema
 * @param {object} options the Handlebars options
 * @returns {string} a string like `1 < x < 2` or `x ≤ 5`
 * @access public
 * @memberOf helpers
 */
function json_schema__number_range (schema, options) {
  let {json_schema__number_range} = versionDispatch[determineVersion(options)]
  return json_schema__number_range(schema)
}

/**
 * Determine the schema-version from the data.
 * The property `$explicitSchemaVersion` of the data object
 * can contain an explicit version.
 * If it does not exist, the version will be derived from the $schema-property.
 * As a fallback, "draft-04" will be expected
 *
 * @param options the Handlebars options-object
 * @returns {string}
 * @access private
 */
function determineVersion (options) {
  if (!options || !options.data || !options.data.root) {
    throw new Error('Expected \'options\' to be a Handlebars options object')
  }
  const root = options.data.root
  return root['$explicitSchemaVersion'] || schemaByUrl[root.$schema] || 'draft-04'
}

/**
 * Compute the element-id for a definition panel
 * @param {string} name the name of the definition (the key within the definitions object)
 * @returns {string} an id without invalid characters for HTML-ids
 * @access public
 * @memberOf helpers
 */
function json_schema__definition_id (name) {
  return `definitions-${htmlId(name)}`
}

/**
 * Returns a prose-description of a number of things (e.g. "at least one item", "2 to 5 items")
 * @param {number=} min the lower bound
 * @param {number=} max the upper bound
 * @param {string} singular the "thing" in it singular form
 * @param {string} plural the "thing" it its plural form
 * @returns {string|null} a prose-description or `null` if there is no range
 * @access public
 * @memberOf helpers
 */
function json_schema__count_range (min, max, singular, plural) {
  if (min == null && max == null) return null
  if (min === 0 && max == null) return null
  if (max === 0) return `no ${plural}`
  if (min === 1 && max == null) return `at least one ${singular}`
  if (min == null && max === 1) return `at most one ${singular}`
  if (min === 1 && max === 1) return `exactly one ${singular}`

  if (min != null && max == null) return `at least ${min} ${plural}`
  if (min == null && max != null) return `at most ${max} ${plural}`
  if (min === max) return `exactly ${max} ${plural}`

  return `${min} to ${max} ${plural}`
}

/**
 * Extract then name of a subschema from a $ref property
 * @param {string} url
 * @returns {*}
 * @access public
 * @memberOf helpers
 */
function json_schema__subschema_name (url) {
  return url.replace('#/definitions/', '')
}

/**
 * Returns "true", if the given propertyName is among the required properties
 * of a schema
 * @param {object} schema the json-schema to check for required properties of
 * @param {string} propertyName the property the name of the property
 * @returns {boolean} true, if the given propertyName is required
 * @access public
 * @memberOf helpers
 */
function json_schema__is_required (schema, propertyName) {
  return schema.required && schema.required.indexOf(propertyName) >= 0
}
/**
 * Return true, if the type is numeric (integer or number) or *could be numeric*
 *
 * The type *could be* numeric, if it is an array that includes number or integer,
 * or if it is not specified.
 *
 * @this {undefined|string|string[]} the json-schema oject
 * @return {boolean} true, if the type could be numeric
 * @access public
 * @memberOf helpers
 */
function json_schema__could_be_numeric () {
  const couldBeA = json_schema__could_be_of_type.bind(this)
  return couldBeA('number') || couldBeA('integer')
}

/**
 * Return true, if the dataType of a schema could be a given type
 *
 * This includes an undefined `type` property, an array `type`-property that includes
 * the checke values and a string `type'-property that equals the checked value.
 *
 * @param {string} type the type to check the schema against
 * @returns {boolean} true, if the type of the schema could be the parameter
 * @access public
 * @memberOf helpers
 */
function json_schema__could_be_of_type (type) {
  const actualType = this.type
  return actualType === null || // type not defined
    actualType === undefined ||
    actualType === type || // explicit check matches
    (Array.isArray(actualType) && actualType.indexOf(type) >= 0) // array contains type
}

/**
 * Splits a coma-separated list into an array.
 *
 * This helper can be used as nested-helper for other helpers that expect arrays
 *
 * @param {string} list a coma-separated list of strings
 * @return {string[]} the list items as array
 * @access public
 * @memberOf helpers
 */
function json_schema__split_coma (list) {
  return list.split(',').map(item => item.trim())
}

/**
 * Returns true, if the value is an array
 * @param {*} value the value
 * @returns {boolean} true, if the value is an array
 * @access public
 * @memberOf helpers
 */
function json_schema__is_array (value) {
  return Array.isArray(value)
}

/**
 * Checks, if the given schema contains any of the provided keywords (like "type" or "items")
 *
 * @param {object} schema a json-schema
 * @param {...string} keywords a list of keywords to check again
 * @returns {boolean} true, if any of the keywords exists in the schema
 * @access public
 * @memberOf helpers
 */
function json_schema__has_any (schema, ...keywords) {
  // remove Handlebars options from keywords before searching
  return !!keywords.slice(0, keywords.length - 1)
    .find(keyword => schema[keyword] != null)
}

/**
 * Convert an array into a prose-enumeration
 * @param {string[]} items an array of strings (e.g. ['a','b','c'])
 * @return {string} a written enumeration (a,b and c)
 * @access public
 * @memberOf helpers
 */
function json_schema__enumerate (items, options) {
  return items.map((item, index, array) => {
    item = options.fn(item)
    // Add coma after each but the two last items
    if (index < array.length - 2) {
      return item + ', '
    }
    // Add "and" after the next-to-last item
    if (index === array.length - 2) {
      return item + ' and '
    }
    return item
  }).join('')
}


handlebars/helpers/utils.js
const Handlebars = require('handlebars')

module.exports = {
  range,
  safe
}

/**
 * Render a range based on several
 * @param {number} min the minimal value or null if not present
 * @param {boolean} minExclusive whether "min" is outside the range
 * @param {number} max the maximal value or null if not present
 * @param {boolean} maxExclusive whether "max" is outside the range
 * @param {string} what a description for the current value
 */
function range (min, minExclusive, max, maxExclusive, what) {
  if (min != null && max != null) {
    return `${min} ${minExclusive ? '<' : '≤'} ${what} ${maxExclusive ? '<' : '≤'} ${max}`
  }
  if (min != null && max == null) {
    return `${what} ${minExclusive ? '>' : '≥'} ${min}`
  }
  if (min == null && max != null) {
    return `${what} ${maxExclusive ? '<' : '≤'} ${max}`
  }
  return null
}

/**
 * Creates a Handlebars.SafeString from a template-literal, escaping the injected values
 * @param strings
 * @param values
 */
function safe (strings, ...values) {
  let escapedValues = values.map(Handlebars.Utils.escapeExpression)
  let rawString = String.raw.apply(this, [strings].concat(escapedValues))
  return new Handlebars.SafeString(rawString)
}


handlebars/helpers/helpers-draft-04.js
/* eslint-disable camelcase */

// draft-04 and draft-05 have the same logic concerning
// minimum, maximum, exclusiveMinimum, exclusiveMaximum
// so we can reuse the helper here
const { json_schema__number_range } = require('./helpers-draft-05')

const metadata = {
  name: 'draft-04',
  docLink: 'https://tools.ietf.org/html/draft-fge-json-schema-validation-00',
  sections: {
    keywords: {
      'multipleOf': '5.1.1',
      'maximum': '5.1.2',
      'exclusiveMaximum': '5.1.2',
      'minimum': '5.1.3',
      'exclusiveMinimum': '5.1.3',
      'maxLength': '5.2.1',
      'minLength': '5.2.2',
      'pattern': '5.2.3',
      'items': '5.3.1',
      'items_array': '5.3.1',
      'additionalItems': '5.3.1',
      'minItems': '5.3.2',
      'maxItems': '5.3.3',
      'uniqueItems': '5.3.4',
      'minProperties': '5.4.2',
      'maxProperties': '5.4.1',
      'required': '5.4.3',
      'properties': '5.4.4',
      'patternProperties': '5.4.4',
      'additionalProperties': '5.4.4',
      'dependencies': '5.4.5',
      'enum': '5.5.1',
      'type': '5.5.2',
      'allOf': '5.5.3',
      'anyOf': '5.5.4',
      'oneOf': '5.5.5',
      'not': '5.5.6',
      'default': '6.2',
      'format': '7'
    },
    formats: {
      'date-time': '7.3.1',
      'email': '7.3.2',
      'hostname': '7.3.3',
      'ipv4': '7.3.4',
      'ipv6': '7.3.5',
      'uri': '7.3.6'
    }
  }
}

module.exports = {
  metadata,
  json_schema__number_range
}


handlebars/helpers/helpers-draft-05.js
/* eslint-disable camelcase */

const {range} = require('./utils')

const metadata = {
  name: 'draft-05',
  docLink: 'https://tools.ietf.org/html/draft-wright-json-schema-validation-00',
  sections: {
    keywords: {
      'multipleOf': '5.1',
      'maximum': '5.2',
      'exclusiveMaximum': '5.3',
      'minimum': '5.4',
      'exclusiveMinimum': '5.5',
      'maxLength': '5.6',
      'minLength': '5.7',
      'pattern': '5.8',
      'items': '5.9',
      'items_array': '5.9',
      'additionalItems': '5.9',
      'maxItems': '5.10',
      'minItems': '5.11',
      'uniqueItems': '5.12',
      'maxProperties': '5.13',
      'minProperties': '5.14',
      'required': '5.15',
      'properties': '5.16',
      'patternProperties': '5.17',
      'additionalProperties': '5.18',
      'dependencies': '5.19',
      'enum': '5.20',
      'type': '5.21',
      'allOf': '5.22',
      'anyOf': '5.23',
      'oneOf': '5.24',
      'not': '5.25',
      'default': '6.2',
      'format': '7'
    },
    formats: {
      'date-time': '7.3.1',
      'email': '7.3.2',
      'hostname': '7.3.3',
      'ipv4': '7.3.4',
      'ipv6': '7.3.5',
      'uri': '7.3.6',
      'uriref': '7.3.7'
    }
  }
}

function json_schema__number_range (schema) {
  return range(schema.minimum, schema.exclusiveMinimum, schema.maximum, schema.exclusiveMaximum, 'x')
}

module.exports = {
  metadata,
  json_schema__number_range
}


handlebars/helpers/helpers-draft-06.js
/* eslint-disable camelcase */

const {range} = require('./utils')

const metadata = {
  name: 'draft-06',
  docLink: 'https://tools.ietf.org/html/draft-wright-json-schema-validation-01',
  sections: {
    keywords: {
      'multipleOf': '6.1',
      'maximum': '6.2',
      'exclusiveMaximum': '6.3',
      'minimum': '6.4',
      'exclusiveMinimum': '6.5',
      'maxLength': '6.6',
      'minLength': '6.7',
      'pattern': '6.8',
      'items': '6.9',
      'items_array': '6.9',
      'additionalItems': '6.10',
      'minItems': '6.11',
      'maxItems': '6.12',
      'uniqueItems': '6.13',
      'contains': '6.14',
      'minProperties': '6.15',
      'maxProperties': '6.16',
      'required': '6.17',
      'properties': '6.18',
      'patternProperties': '6.19',
      'additionalProperties': '6.20',
      'dependencies': '6.21',
      'propertyNames': '6.22',
      'enum': '6.23',
      'const': '6.24',
      'type': '6.25',
      'allOf': '6.26',
      'anyOf': '6.27',
      'oneOf': '6.28',
      'not': '6.29',
      'default': '7.3',
      'examples': '7.4',
      'format': '8'
    },
    formats: {
      'date-time': '8.3.1',
      'email': '8.3.2',
      'hostname': '8.3.3',
      'ipv4': '8.3.4',
      'ipv6': '8.3.5',
      'uri': '8.3.6',
      'uri-reference': '8.3.7',
      'uri-template': '8.3.8',
      'json-pointer': '8.3.9'
    }
  }
}

/**
 *
 * @param {object} schema a json-schema with numeric bound restrictions
 * @param {number=} schema.exclusiveMinimum an exclusive minimum boundary
 * @param {number=} schema.exclusiveMaximum an exclusive minimum boundary
 * @param {number=} schema.minimum an inclusive minimum boundary
 * @param {number=} schema.maximum an inclusive minimum boundary
 * @returns {*}
 */
function json_schema__number_range (schema) {
  // The actual minimum boundary is exclusive, if "minimum" either does not exist or is larger than "exclusiveMinimum"
  let minExclusive = schema.exclusiveMinimum != null && (schema.minimum == null || schema.exclusiveMinimum >= schema.minimum)

  // The actual maximum boundary is exclusive, if "maximum" either does not exist or is smaller than "exclusiveMaximum"
  let maxExclusive = schema.exclusiveMaximum != null && (schema.maximum == null || schema.exclusiveMaximum <= schema.maximum)

  return range(
    minExclusive ? schema.exclusiveMinimum : schema.minimum,
    minExclusive,
    maxExclusive ? schema.exclusiveMaximum : schema.maximum,
    maxExclusive,
    'x')
}

module.exports = {
  metadata,
  json_schema__number_range
}
